source/sliding_puzzle.py

```diff
@@ -124,7 +139,63 @@ def _join(self, row: List[PuzzlePiece]) -> List[str]:
     def draw(self) -> str:
         """Draw the puzzle in its current state."""
         output = []
-        for row in self.pieces:
+        for row in self.rows:
             output.extend(self._join(row))
 
         return '\n'.join(output)
+
+    def shuffle(self) -> None:
+        """Shuffle the puzzle by randomizing the order of the pieces"""
+        for row in self.rows:
+            random.shuffle(row)
+        random.shuffle(self.rows)
+        # empty one of the rows
+        random.choice(random.choice(self.rows)).clear()
+
+    def _get_empty_piece_position(self) -> PiecePosition:
+        x, y, index = [
+            (x, y, piece.index)
+            for x, row in enumerate(self.rows)
+            for y, piece in enumerate(row)
+            if piece.empty
+        ][0]
+        return PiecePosition(x=x, y=y, index=index)
+
+    def move_up(self) -> None:
+        """Move the piece above the empty piece down"""
+        empty_pos = self._get_empty_piece_position()
+        # return if empty piece is on the first row
+        if empty_pos.x == 0:
+            return
+        self._swap_pieces(x1=empty_pos.x, y1=empty_pos.y, x2=empty_pos.x - 1, y2=empty_pos.y)
+
+    def move_down(self) -> None:
+        """Move the piece below the empty piece up"""
+        empty_pos = self._get_empty_piece_position()
+        # return if empty piece is on the last row
+        last_row_x = len(self.rows[0]) - 1
+        if empty_pos.x == last_row_x:
```

Comment:
You never use the last row variable again
```suggestion
        if empty_pos.x == len(self.rows[0]) - 1:
```